x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SearchableSnapshotAction.java

```diff
@@ -0,0 +1,203 @@
+        StepKey copyMetadataKey = new StepKey(phase, NAME, CopyExecutionStateStep.NAME);
+        StepKey copyLifecyclePolicySettingKey = new StepKey(phase, NAME, CopySettingsStep.NAME);
+        StepKey swapAliasesKey = new StepKey(phase, NAME, SwapAliasesAndDeleteSourceIndexStep.NAME);
+
+        WaitForNoFollowersStep waitForNoFollowersStep = new WaitForNoFollowersStep(waitForNoFollowerStepKey, generateSnapshotNameKey,
+            client);
+        GenerateSnapshotNameStep generateSnapshotNameStep = new GenerateSnapshotNameStep(generateSnapshotNameKey, cleanSnapshotKey);
+        CleanupSnapshotStep cleanupSnapshotStep = new CleanupSnapshotStep(cleanSnapshotKey, takeSnapshotKey, client, snapshotRepository);
+        TakeSnapshotStep takeSnapshotStep = new TakeSnapshotStep(takeSnapshotKey, waitForSnapshotInProgressKey, client, snapshotRepository);
+        WaitForSnapshotInProgressStep waitForSnapshotInProgressStep = new WaitForSnapshotInProgressStep(waitForSnapshotInProgressKey,
+            verifySnapshotStatusBranchingKey, snapshotRepository);
+        OnAsyncWaitBranchingStep onAsyncWaitBranchingStep = new OnAsyncWaitBranchingStep(verifySnapshotStatusBranchingKey,
+            cleanSnapshotKey, restoreFromSearchableRepoKey, client, getCheckSnapshotStatusAsyncAction(snapshotRepository));
+        RestoreSnapshotStep restoreSnapshotStep = new RestoreSnapshotStep(restoreFromSearchableRepoKey, waitForGreenRestoredIndexKey,
+            client, searchableRepository, RESTORED_INDEX_PREFIX);
+        WaitForGreenIndexHealthStep waitForGreenIndexHealthStep = new WaitForGreenIndexHealthStep(waitForGreenRestoredIndexKey,
+            copyMetadataKey, RESTORED_INDEX_PREFIX);
+        CopyExecutionStateStep copyMetadataStep = new CopyExecutionStateStep(copyMetadataKey, copyLifecyclePolicySettingKey,
+            RESTORED_INDEX_PREFIX, nextStepKey.getName());
+        CopySettingsStep copySettingsStep = new CopySettingsStep(copyLifecyclePolicySettingKey, swapAliasesKey, client,
+            RESTORED_INDEX_PREFIX, LifecycleSettings.LIFECYCLE_NAME);
+        // sending this step to null as the restored index (which will after this step essentially be the source index) was sent to the next
+        // key after we restored the lifecycle execution state
+        SwapAliasesAndDeleteSourceIndexStep swapAliasesAndDeleteSourceIndexStep = new SwapAliasesAndDeleteSourceIndexStep(swapAliasesKey,
+            null, client, RESTORED_INDEX_PREFIX);
+
+        return Arrays.asList(waitForNoFollowersStep, generateSnapshotNameStep, cleanupSnapshotStep, takeSnapshotStep,
+            waitForSnapshotInProgressStep, onAsyncWaitBranchingStep, restoreSnapshotStep, waitForGreenIndexHealthStep,
+            copyMetadataStep, copySettingsStep, swapAliasesAndDeleteSourceIndexStep);
+    }
+
+    private TriConsumer<Client, IndexMetaData, AsyncWaitStep.Listener> getCheckSnapshotStatusAsyncAction(String snapshotRepository) {
```

Comment:
Can you add javadocs to this method?

Reply:
And maybe this can be static? We could pass in the names of things, might make it easier to test?